src/interceptors/ClientRequest/MockHttpSocket.ts

```diff
@@ -0,0 +1,450 @@
+
+        if (done) {
+          break
+        }
+
+        // Flush the headers upon the first chunk in the stream.
+        // This ensures the consumer will start receiving the response
+        // as it streams in (subsequent chunks are pushed).
+        if (httpHeaders.length > 0) {
+          flushHeaders(value)
+          continue
+        }
+
+        // Subsequent body chukns are push to the stream.
+        this.push(value)
+      }
+    }
+
+    // If the headers were not flushed up to this point,
+    // this means the response either had no body or had
+    // an empty body stream. Flush the headers.
+    if (httpHeaders.length > 0) {
+      flushHeaders()
+    }
+
+    // Close the socket if the connection wasn't marked as keep-alive.
+    if (!this.shouldKeepAlive) {
+      this.emit('readable')
+
+      /**
+       * @todo @fixme This is likely a hack.
+       * Since we push null to the socket, it never propagates to the
+       * parser, and the parser never calls "onResponseEnd" to close
+       * the response stream. We are closing the stream here manually
+       * but that shouldn't be the case.
+       */
+      this.responseStream?.push(null)
+      this.push(null)
+    }
+  }
+
+  /**
+   * Close this Socket connection with the given error.
+   */
+  public errorWith(error: Error): void {
+    this.destroy(error)
+  }
+
+  private mockConnect(): void {
+    this.emit('lookup', null, '::1', 6, this.connectionOptions.host)
+    this.emit('connect')
+    this.emit('ready')
+
+    if (this.baseUrl.protocol === 'https:') {
+      this.emit('secure')
+      this.emit('secureConnect')
+      // A single TLS connection is represented by two "session" events.
+      this.emit('session', Buffer.from('mock-session-renegotiate'))
+      this.emit('session', Buffer.from('mock-session-resume'))
+    }
+  }
+
+  private onRequestStart: RequestHeadersCompleteCallback = (
+    versionMajor,
+    versionMinor,
+    rawHeaders,
+    _,
+    path,
+    __,
+    ___,
+    ____,
+    shouldKeepAlive
+  ) => {
+    this.shouldKeepAlive = shouldKeepAlive
+
+    const url = new URL(path, this.baseUrl)
+    const method = this.connectionOptions.method || 'GET'
+    const headers = parseRawHeaders(rawHeaders)
+    const canHaveBody = method !== 'GET' && method !== 'HEAD'
+
+    // Translate the basic authorization in the URL to the request header.
+    // Constructing a Request instance with a URL containing auth is no-op.
+    if (url.username || url.password) {
+      if (!headers.has('authorization')) {
+        headers.set('authorization', `Basic ${url.username}:${url.password}`)
+      }
+      url.username = ''
+      url.password = ''
+    }
+
+    // Create a new stream for each request.
+    // If this Socket is reused for multiple requests,
+    // this ensures that each request gets its own stream.
+    // One Socket instance can only handle one request at a time.
+    if (canHaveBody) {
+      this.requestStream = new Readable()
```

Comment:
Don't we need to implement the `read` function here?

I get:
> Error [ERR_METHOD_NOT_IMPLEMENTED]: The _read() method is not implemented
    at Readable._read (node:internal/streams/readable:904:9)
    at Readable.read (node:internal/streams/readable:737:12)
    at maybeReadMore_ (node:internal/streams/readable:891:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:

When trying to use stream request:
```js
const http = require('http')
const { ClientRequestInterceptor } = require('@mswjs/interceptors/ClientRequest')
const { Readable } = require('stream');

const interceptor = new ClientRequestInterceptor()

interceptor.on('request', async function rootListener({ request }) {
  console.log(await request.arrayBuffer());
  request.respondWith(new Response('OK!'))
});
interceptor.apply();

const req = http.request('http://webhook.site/b1c9419d-4706-4f1c-9039-6825cfa5e77f', { method: 'POST' }, res => {
  console.log(res.statusCode)
})
let counter = 0
const r = new Readable({
  read: () => {
    if (counter++ < 3) {
      r.push('dsa1')
    } else {
      r.push(null)
    }
  }
})

r.pipe(req)

setTimeout(req.end.bind(req), 300)
```

Reply:
Added in #527